(Closes #2094) Small fix to signature access to handle firstprivate structures

LonelyCat124 opened the pull request:
See #2094.

hiker commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -705,7 +705,10 @@ def _get_private_clauses(self):
                         # The assignment to the variable is inside a loop, so
                         # declare it to be private
-                        name = str(signature).lower()
+                        if str(signature).find("%") > 0:
```
I just noticed this: this is the same as `signature[0].lower()` (unless I am missing something)

LonelyCat124 replied:
Ok I'll swap it to that.

arporter commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -113,6 +114,72 @@ def test_ompparallel_changes_begin_string(fortran_reader):
 
 
+def test_ompparallel_structures_private_clause():
+    '''TODO'''
```
Please do :-)

LonelyCat124 replied:
Added a description.

arporter commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -113,6 +114,72 @@ def test_ompparallel_changes_begin_string(fortran_reader):
 
+def test_ompparallel_structures_private_clause():
+    '''TODO'''
+    subroutine = Routine("testsub")
```
While I'm impressed that you've constructed all the PSyIR for this by hand, I think your life would be much simpler (and the code shorter) if you 'cheated' and just used the `fortran_reader` test fixture with a suitable bit of Fortran.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -705,7 +705,7 @@ def _get_private_clauses(self):
                     if loop_ancestor:
                         # The assignment to the variable is inside a loop, so
                         # declare it to be private
```
Please extend the comment. Something like "...to be private. Ensure we only use the root variable name in case this is an access of a structure member (see 3.2.1 of the OMP spec. v.5.2)."

LonelyCat124 replied:
Added

arporter commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -113,6 +114,72 @@ def test_ompparallel_changes_begin_string(fortran_reader):
+    subroutine = Routine("testsub")
+    parallel = OMPParallelDirective.create()
+    subroutine.addchild(parallel)
+    sing = OMPSingleDirective()
```
pylint complains that `sing` is unused.

LonelyCat124 replied:
Fixed by redoing the test to use fortran_reader

arporter reviewed (changes requested):
Fix is good. I think the test can be shortened somewhat and then this is good to go.

arporter commented:
HI @LonelyCat124, have you had a chance to look at this one again yet?

LonelyCat124 commented:
I don't even remember doing this somehow, I will address comments now.

LonelyCat124 commented:
I've addressed the comments now - sorry for the delay. 

One question I have though now about this PR is to @sergisiso is this handled by #2124 anyway - in which case we can just wait for that PR perhaps instead?

sergisiso commented:
Yes I do the same on that PR.

sergisiso commented:
It won't close the issue because doing `signature[0]` (same as this PR) it just means we get the top-level reference, which is syntactically right but in some cases we need more granularity if different parts of the struct should have different attributes. (specially in requesting synchonisations because any read-write to any part of the struct is considered a race condition).  But I think it solves the cases we found for now.

LonelyCat124 commented:
I realised now that this may produce a bug in a case where one element of a structure is deemed to be `private` and another is deemed to be `firstprivate`, resulting in the stucture appearing in both clauses which I believe is forbidden, I'll try to have a think to fix this too when I have time

arporter commented:
So shall I proceed to review/merge #2124 for now?

sergisiso commented:
https://github.com/stfc/PSyclone/pull/2124 is on master, I thin this PR can be closed now (not the issue yet unfortunately )

arporter commented:
Closing as fix is now on master.
